File: Extras/pytest/unium.py
import pytest
import json
import async_timeout

class WebsocketHelper:

  """A class to make the websocket interface more test friendly"""

  def __init__( self, ws ):
    """instantiate with the websocket to use"""
    self.ws = ws
    self.next_id = 1
    self._on = {}
    self._once = {}


  def _handle_message( self, msg ):
    """fire off any event handlers waiting on the given message"""

    if not "id" in msg:
      return

    mid = msg["id"]

    if mid in self._on:
      for fn in self._on[ mid ]:
        fn( msg )

    if mid in self._once:
      for fn in self._once[ mid ]:
        fn( msg )
      del self._once[ mid ]


  async def wait_for( self, event, timeout=2.0 ):
    """wait until we receive a message with a specific id (or timeout)"""

    async with async_timeout.timeout( timeout ):
      while True:

        msg = json.loads( await self.ws.recv() )
        self._handle_message( msg )

        if "id" in msg and msg["id"] == event:
          return msg


  async def send( self, uri, name="" ):
    """send a message, return message id"""

    if name == "":
      mid = f"m{self.next_id}"
      self.next_id += 1
    else:
      mid = name

    await self.ws.send( json.dumps( { 'id': mid, 'q' : uri } ))

    return mid


  async def get( self, uri, timeout=1.0 ):
    """send a message and return the response (or timeout)"""

    async with async_timeout.timeout( timeout ):

      mid = await self.send( uri )
      msg = await self.wait_for( mid, timeout )

      assert "id" in msg
      assert "data" in msg
      assert msg["id"] == mid

      return msg['data']


  async def bind( self, uri, name="" ):
    """
    register to receive events
    uses the event name from the URI as the identifier unless we've given given one explicitly
    """
    mid = name if name else uri.split(".")[-1]

    await self.send( f"/bind{uri}", mid )

    response = await self.wait_for( mid )
    assert "id" in response
    assert "info" in response

    return response


  def on( self, event, function ):
    """register an event handler - will be fired every time a message with a given id is received"""
    if not event in self._on:
      self._on[ event ] = []
    self._on[ event ].append( function )


  def once( self, event, function ):
    """register an event handler - will be fired once a message with a given id is received"""
    if not event in self._once:
      self._once[ event ] = []
    self._once[ event ].append( function )

File: Extras/cucumber.js/features/step_definitions/unium.js
// a class to make the websocket interface a little more test friendly
// by providing a promise/async version of the websocket interface

const chai      = require( 'chai' );
const WebSocket = require( 'ws' );

const { expect } = chai;

module.exports = class UniumHelper {

  constructor() {
    this.ws      = null;
    this.next_id = 1;
    this._on     = {};
    this._once   = {};
  }

  // async connect

  connect( uri ) {
    return new Promise( (resolve, reject) => {
      this.disconnect();
      this.ws = new WebSocket( uri );
      this.ws.on( 'open', resolve );
      this.ws.on( 'error', reject );
      this.ws.on( 'message', (m) => this.onMessage(m) );
    });
  }

  disconnect() {
    if( this.ws ) {
      this.ws.close();
      this.ws = null;
    }
  }


  // when a message is received, invoke any handlers registered for this event

  onMessage( msg ) {

    const data = JSON.parse( msg );
    const {id} = data;

    if(!id) {
      return;
    }

    if( id in this._on ) {
      this._on[ id ].forEach( fn => fn( data ) );
    }

    if( id in this._once ) {
      this._once[ id ].forEach( fn => fn( data ) );
      delete this._once[ id ];
    }
  }

  // add event handlers

  on( event, fn ) {
    if( event in this._on ) {
      this._on[ event ].push( fn );
    } else {
      this._on[ event ] = [ fn ];
    }
  }

  once( event, fn ) {
    if( event in this._once ) {
      this._once[ event ].push( fn );
    } else {
      this._once[ event ] = [ fn ];
    }
  }

  // send message, return message id

  send( uri, id ) {
    const msg = {
      id: id || `m${this.next_id++}`,
      q: uri
    };
    this.ws.send( JSON.stringify( msg ) )
    return msg.id;
  }

  // wait until we receive a message with a given id
  
  wait_for( id, timeout = 2 ) {
    return new Promise( (resolve,reject) => {
      const timeoutHandle = setTimeout( () => reject( 'timeout out' ), timeout * 1000 );
      this.once( id, (data) => {
        clearTimeout( timeoutHandle );
        resolve( data );
      });
    });
  }

  // send a query and result the result (like a HTTP get)

  async get( uri, timeout = 2 ) {

    const id  = await this.send( uri );
    const res = await this.wait_for( id, timeout );

    expect( res ).to.not.be.null;
    res.should.have.property( 'id' );
    res.id.should.equal( id );
    res.should.not.have.property( 'error' );
    res.should.have.property( 'data' );

    return res.data;
  }
  
  // convenient function to binding to events
  // use event name as id unless given on

  async bind( uri, name ) {

    const id = name || uri.split('.').pop();

    await this.send( `/bind${uri}`, id );
    const res = await this.wait_for( id );

    expect( res ).to.not.be.null;
    res.should.have.property( 'id' );
    res.should.not.have.property( 'error' );
    res.should.have.property( 'info' );

    return res;
  }
}
